### PR TITLE
Restructure AHRS in preparation for frontend/backend split

### PR DESCRIPTION
**Pull request by peterbarker**
This renames `AP_AHRS.*` to `AP_AHRS_Backend.*` and then `AP_AHRS_NavEKF.*` to `AP_AHRS.*`, and renames `AP_AHRS` -> `AP_AHRS_Backend` and then `AP_AHRS_NavEKF` -> `AP_AHRS`

Small adjustments were made for things like moving parameters to the new `AP_AHRS`.

There's now just AP_AHRS - no AP_AHRS_NavEKF object or AP::ahrs_navekf() accessor.

It also stops `AP_Vehicle` from giving the option to instantiate an `AP_AHRS_DCM` - on the assumption that the new `AP_AHRS` will cut all irrelevant stuff out based on `HAL_NAVEKF2_ENABLED` etc.

Rough plan for future PR making the split:
  - make `AP_AHRS_DCM` an actual backend
    - `AP_AHRS` stops inheritting from `AP_AHRS_DCM` and instead encapsulates an instance of `AP_AHRS_DCM`
      - all methods have `override` removed.  Where we were relying on behaviour from `AP_AHRS_DCM` or `AP_AHRS_Backend` we call a method on the `AP_AHRS_DCM _dcm;` object
      - `AP_AHRS_DCM`'s `get_position` method loses `virtual`.  The frontend will need smarts for this method, or we add a `allow_dead_reckoning` `bool` to `get_location` instead
  - make SITL an actual `AP_AHRS_Backend`
  - make External an actual `AP_AHRS_Backend`
  - make `NavEKF2` and `NavEKF3` backends
  - change all switch statements to work on an `*_active_ekf`
  - add a _backends array to iterate over, use it

Structure:
image

### CODE REVIEW

**Comment by peterbarker** [comment]
Note that this patch really isn't actually all that big.  The renames aren't being handled well.

**Comment by peterbarker** [comment]
```
                     master: bin/arduplane  1595548  2436  194380  1792364
                    This PR: bin/arduplane  1595560  2436  194376  1792372
 master / no EKF*_AVAILABLE: bin/arduplane  1350964  2432  194380  1547776
This PR / no EKF*_AVAILABLE: bin/arduplane  1350960  2432  194376  1547768
```

**Comment by peterbarker** [comment]
I removed the `AP_AHRS_NAVEKF_AVAILABLE` with this PR.  I thought I'd do a build with that set to zero to see what size difference that made.  It doesn't compile, and not by a small amount for Plane (quadplane...).  Rover also has tie-ins to the EKF which make compiling it out difficult (ekf_checks).

The exercise did show that removing `AP_AHRS_NAVEKF_AVAILABLE` isn't a terrible thing.

It also showed that `get_filter_status` should probably be pure-virtual on the new `AP_AHRS_Backend` class and implemented by DCM.

File: ArduCopter/AP_Arming.cpp [hunk 1]
```diff
@@ -512,7 +512,7 @@ bool AP_Arming_Copter::mandatory_gps_checks(bool display_failure)
     bool mode_requires_gps = copter.flightmode->requires_GPS();
 
     // always check if inertial nav has started and is ready
-    const AP_AHRS_NavEKF &ahrs = AP::ahrs_navekf();
+    const AP_AHRS &ahrs = AP::ahrs();
```
Comment (tridge) [inline ArduCopter/AP_Arming.cpp:515]: i'd prefer "const auto &ahrs = AP::ahrs();"

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduCopter/AP_Arming.cpp (hunk 1 above)
Comment (peterbarker) [reply ArduCopter/AP_Arming.cpp]: Done.

**Comment by peterbarker** [comment]
It was agreed on the call this could be merged once the `auto` keyword was used and it passed CI.

**Pull request merged**